src/main/java/net/glowstone/constants/GlowParticle.java

```diff
@@ -71,58 +76,139 @@ private GlowParticle() {
-                }
                 return EMPTY;
         }
     }
 
     /**
+     * Convert an object to a data array if possible for a particle.
+     *
+     * @param particle the Particle to validate.
+     * @param object the Object to convert.
+     * @return The data array for the particle effect.
+     */
+    public static int[] getData(Particle particle, Object object) {
+        if (particle.getDataType() == Void.class) {
+            return new int[0];
+        }
+
+        if (particle.getDataType() == MaterialData.class) {
+            if (object == null) {
+                return new int[]{0};
+            }
+
+            MaterialData material = (MaterialData) object;
+
+            return new int[]{material.getItemTypeId() + (material.getData() << 12)};
+        }
+
+        if (particle.getDataType() == ItemStack.class) {
+            if (object == null) {
+                return TWO_EMPTY;
+            }
+
+            ItemStack item = (ItemStack) object;
+
+            // http://wiki.vg/Protocol#Particle
+            // data "Length depends on particle. "iconcrack" [Effect.ITEM_BREAK] has length of 2, "blockcrack",
+            // and "blockdust" have lengths of 1, the rest have 0"
+            // iconcrack_(id)_(data) 36
+            return new int[]{item.getTypeId(), item.getDurability()};
+        }
+
+        // doesn't make sense, there are only 3 particle data types
+        return null;
+    }
+
+
+    /**
      * Determine whether a particle type is considered long distance, meaning
      * it has a higher visible range than normal.
      *
      * @param particle the Particle.
      * @return True if the particle is long distance.
      */
     public static boolean isLongDistance(Effect particle) {
-        return particle == Effect.EXPLOSION ||
-                particle == Effect.EXPLOSION_LARGE ||
-                particle == Effect.EXPLOSION_HUGE ||
-                particle == Effect.MOB_APPEARANCE;
+        switch (particle) {
+            case EXPLOSION:
+            case EXPLOSION_LARGE:
+            case EXPLOSION_HUGE:
+            case MOB_APPEARANCE:
+                return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * Determine whether a particle type is considered long distance, meaning
+     * it has a higher visible range than normal.
+     *
+     * @param particle the Particle.
+     * @return True if the particle is long distance.
+     */
+    public static boolean isLongDistance(Particle particle) {
+        switch (particle) {
+            case EXPLOSION_NORMAL:
+            case EXPLOSION_LARGE:
+            case EXPLOSION_HUGE:
+            case MOB_APPEARANCE:
+                return true;
+        }
+
+        return false;
     }
 
-    private static void set(Effect particle, int id) {
-        ids[particle.ordinal()] = id;
+    private static void set(Particle particle, Effect effect, int id) {
```

Comment:
I thought since you made it a `HashMap`, you wanted to support mods adding new particle effects.